### ISSUE TITLE
Permissions of removed users and groups stay in database

### ISSUE THREAD

**Issue by jefscheepers** [irods/irods#7778]
- [x] main
- [x] 4-3-stable

---

### iRODS Version, OS and Version

iRODS version 4.3.1, AlmaLinux 8.8
MySQL 8 as database.
iCommands version 4.3.1

### What did you try to do?

I gave a group and a user permissions on a collection.
I removed both the group and the user from the zone, and then queried for the permissions on the collection.

### Expected behavior

The permissions for the removed group and user are removed.

### Observed behavior (including steps to reproduce, if applicable)

I cannot see the permissions for the removed group/user with ils -A, but I can still see them with iquest or with session.query in Python.
I can see the ID, but I cannot find a username for the ID.
It seems to me like the link between the user ID, access right and collection are still kept somewhere in the database.
Since the group doesn't exist anymore, I don't believe that there is a real access issue. However, this outdated info remains, which in my opinion is confusing and requires some manual cleaning after queries.
These results can be seen both by rodsadmins and rodsusers.

### Steps to reproduce (as rodsadmin):

1. Setup: creating the collection, group, user and permissions

```
$ imkdir test
$ iadmin mkgroup testgroup
$ iadmin mkuser bob rodsuser
$ iadmin atg testgroup bob
$ ichmod read testgroup test
$ ichmod write bob test
```

2. querying for permissions with ils and iquest

```
$ ils -A test
```

> /<zone>/home/rods/test:
>         ACL - rods#<zone>:own   g:testgroup#<zone>:read_object   bob#<zone>:modify_object
>         Inheritance - Disabled

```
$ iquest "SELECT COLL_ACCESS_USER_ID, COLL_ACCESS_NAME where COLL_NAME = '/<zone>/home/rods/test'"
```

> COLL_ACCESS_USER_ID = 10002
> COLL_ACCESS_NAME = own
> 
> COLL_ACCESS_USER_ID = 10020
> COLL_ACCESS_NAME = read_object
> 
> COLL_ACCESS_USER_ID = 10021
> COLL_ACCESS_NAME = modify_object
> 

3. removing the user and group

```
$ iadmin rmgroup testgroup
$ iadmin rmuser bob
```

5. Repeat step 2

```
$ ils -A test
```

> /<zone>/home/rods/test:
>         ACL - rods#<zone>:own
>         Inheritance - Disabled

```
$ iquest "SELECT COLL_ACCESS_USER_ID, COLL_ACCESS_NAME where COLL_NAME = '/<zone>/home/rods/test'"
```

> COLL_ACCESS_USER_ID = 10002
> COLL_ACCESS_NAME = own
> 
> COLL_ACCESS_USER_ID = 10020
> COLL_ACCESS_NAME = read_object
> 
> COLL_ACCESS_USER_ID = 10021
> COLL_ACCESS_NAME = modify_object
> 

As you see, the results for iquest are the same.

**Comment by mstfdkmn** [comment]
And when a said user is recreated, and it is given an access (or the same access type) on the said collection, the query result shows 2 different ids for that collection (I guess because a new user record (for the same user name) has the new id).

**Comment by trel** [comment]
> And when a said user is recreated, and it is given an access (or the same access type) on the said collection, the query result shows 2 different ids for that collection (I guess because a new user record (for the same user name) has the new id).

That is correct and expected.   And at this time, there is no leakage of permissions as the original user is still 'gone'.

**Comment by mstfdkmn** [comment]
Correct, no leakage of permissions. Just wanted to highlight a workaround need for a simple use case like below. Because we can query acls with `COLL_ACCESS_USER_ID` via the qenquery but we give accesses to users with their user_name (different than how the PRC works - it has its own query). Say I have a policy like;
- get all available permissions on a collection
- store them as metadata on teh said collection
- change the access type of users/groups that exists on the said collection
- some time later (after a logic is implemented) restore the original permission level (querying metadata, and querying user_name based on user_id - fetched from metadata query) 

So in the current situation I see confusing results in case there are removed and recreated users. But okay we add some checks.

**Comment by trel** [comment]
Yes, we should definitely clean up the removed users more completely.
